### PR TITLE
add mitre attack based auto-correlations support in correlation engine

### PR DESCRIPTION
**Pull request by sbcd90**
### Description
Findings in security-analytics link sigma rules to logs. Each sigma rule is linked to one or more mitre tactics & techniques. this info along with mitre cti relationships are used to generate auto-correlations among findings.
this pr add mitre attack based auto-correlations support in correlation engine. https://github.com/mitre/cti

image

 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by eirsep** [comment]
plz add more elaborate description

File: src/main/resources/correlations/mitre_correlation.json [hunk 1]
```diff
@@ -0,0 +1,9625 @@
+{
```
Comment (eirsep) [inline src/main/resources/correlations/mitre_correlation.json:1]: how can we verify this

Comment (sbcd90) [reply src/main/resources/correlations/mitre_correlation.json:1]: we generated this from https://github.com/mitre/cti

**Review by praveensameneni — APPROVED** [review]

File: src/main/java/org/opensearch/securityanalytics/correlation/JoinEngine.java [hunk 2]
```diff
@@ -58,18 +65,147 @@ public class JoinEngine {
...
         this.corrTimeWindow = corrTimeWindow;
         this.correlateFindingAction = correlateFindingAction;
+        this.logTypeService = logTypeService;
     }
 
     public void onSearchDetectorResponse(Detector detector, Finding finding) {
+        try {
+            generateAutoCorrelations(detector, finding);
+        } catch (IOException ex) {
+            correlateFindingAction.onFailures(ex);
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/correlation/JoinEngine.java:87]: can we log error message that auto correlation has failed

File: src/main/java/org/opensearch/securityanalytics/correlation/JoinEngine.java [hunk 3]
```diff
@@ -58,18 +65,147 @@ public class JoinEngine {
...
+                    client.multiSearch(mSearchRequest, new ActionListener<>() {
+                        @Override
+                        public void onResponse(MultiSearchResponse items) {
+                            MultiSearchResponse.Item[] responses = items.getResponses();
+
+                            Map<String, List<String>> autoCorrelationsMap = new HashMap<>();
+                            int idx = 0;
+                            for (MultiSearchResponse.Item response : responses) {
+                                if (response.isFailure()) {
+                                    log.info(response.getFailureMessage());
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/correlation/JoinEngine.java:145]: why info log? error or debug would be better. along with description message of what this failure is from?

**Review by eirsep — APPROVED** [review]
I am approving this
but can we take a follow up action item to add code comments.
This will make the repo and this feature more maintainable and understandable to community

**Comment by sbcd90** [comment]
updated to a more verbose description. added review comments to issue #502

**Pull request merged**